source/manual/conventions-in-rails-applications.html.md

```diff
@@ -0,0 +1,301 @@
+---
+owner_slack: "#govuk-developers"
+title: Conventions for Rails applications
+section: Applications
+type: learn
+layout: manual_layout
+parent: "/manual.html"
+last_reviewed_on: 2020-05-26
+review_in: 12 months
+---
+
+The majority of GOV.UK applications are built using [Ruby on Rails][] and over
+the years we have created a lot them. Along the way we've used a variety of
+patterns and conventions. We aim for GOV.UK apps to be [consistent in their
+implementation][consistent-govuk] with differences in domain logic. However,
+it isn't practical to evolve all apps at the same pace so there can be
+differences in approaches used.
+
+This document serves as a guide to the _current_ conventions and preferences. It
+is intended to serve as a guide for teams creating new applications and for
+teams iterating older ones. If it is out of date, please update it.
+
+Non-Rails GOV.UK applications may still benefit from these conventions.
+Consider applying them they do not conflict with existing
+conventions in an adopted framework.
+
+[Ruby on Rails]: https://rubyonrails.org/
+[consistent-govuk]: https://docs.google.com/document/d/1jEjPOFJ2s1cjQv9vHAbE-KF68LSletpnUVaG5lXlHy4/edit#heading=h.yod40rauhyhu
+
+## Follow industry conventions
+
+GOV.UK's use of Rails is intended to be consistent with the conventions of
+the framework itself and we intend to embrace developments and deprecations
+from the framework. Where GOV.UK apps differ from Rails conventions (such as
+using the RSpec testing framework) these are intended to be consistent with
+common industry practices.
```

Comment:
I guess when I read this, maybe you mean something more like "best practices", or actually common practice, which I guess means the same as popular practice, so do it how most other people do. Which was more your intent?

Personally, my view is that using tools in accordance with their affordances is beneficial, but it's encouraged to take different approaches when it's discussed and supported by some reasoning/evidence.

Taking the example of RSpec further, that's not the Rails default, and Basecamp, who I would see as setting industry conventions with regards to Rails (as they're the main industry body related to it) don't use RSpec as far as I'm aware, but again I'm personally all for making this less about what others use, and more about what the people working on GOV.UK think would be better for GOV.UK.

Reply:
Thanks for commenting. 👍 

I guess I'm hoping for common and good practices. I'm not sure that outside of Rails' own guides there are clear best practice, so forgoing that I'd suggest things that are commonly done that, collectively, people tend to agree are good (or at least not anti-patterns).

There are some things that may produce a really good result, such as our (internally controversial) use of hexagonal architecture, that in some circles are a best practice but, because it's not a particularly common approach, can be surprising. This can make it difficult for developers familiar with Rails' conventions to learn those applications. 

I think the advantage in being somewhat consistent with Basecamp and wider Rails industry is that we're continuing in the same stream as others and thus are more likely to find our tooling supported, maintained and compatible with future developments.